File: refsol/combination-sum.py
def combinationSum(candidates, target):
    res = []
    candidates.sort()

    def backtrack(start, remaining, path):
        if remaining == 0:
            res.append(path[:])
            return
        for i in range(start, len(candidates)):
            c = candidates[i]
            if c > remaining:
                break
            path.append(c)
            backtrack(i, remaining - c, path)
            path.pop()

    backtrack(0, target, [])
    return res

File: refsol/combination-sum.js
export function combinationSum(candidates, target) {
  const res = [];
  candidates.sort((a, b) => a - b);

  function backtrack(start, remaining, path) {
    if (remaining === 0) {
      res.push(path.slice());
      return;
    }
    for (let i = start; i < candidates.length; i++) {
      const c = candidates[i];
      if (c > remaining) break;
      path.push(c);
      backtrack(i, remaining - c, path);
      path.pop();
    }
  }

  backtrack(0, target, []);
  return res;
}
